Realm/RLMResults.mm

```diff
@@ -244,6 +244,24 @@ - (NSUInteger)indexOfObject:(RLMObject *)object {
     return result;
 }
 
+- (id)valueForKey:(NSString *)key {
+    RLMResultsValidate(self);
+    const size_t size = _backingView.size();
+    __block size_t index = 0;
+    return RLMCollectionValueForKey(key, _realm, _objectSchema, size, ^size_t{
+        return _backingView.get(index++).get_index();
+    });
+}
+
+- (void)setValue:(id)value forKey:(NSString *)key {
+    RLMResultsValidateInWriteTransaction(self);
+    const size_t size = _backingView.size();
+    __block size_t index = 0;
+    RLMCollectionSetValueForKey(value, key, _realm, _objectSchema, size, ^size_t{
+        return _backingView.get(index++).get_index();
```

Comment:
`_backingView.get_source_ndx(index++)`

Also maybe just add `indexInSource` to RLMArrayLinkView, expose it in the _Private headers, and use that rather than the block.